### PR TITLE
Fix try_number handling when db isolation enabled

### PR DESCRIPTION
**Pull request by dstandish**
There was an error in the refresh_from_db code, and because of try_number inconsistency, the same run was going into two different log files.  There is some ugliness here, but some ugliness is unavoidable when dealing with try_number as it is right now.

### CODE REVIEW

File: airflow/models/dagrun.py [hunk 1]
```diff
@@ -650,7 +650,6 @@ def get_task_instance(
         )
 
     @staticmethod
-    @internal_api_call
```
Comment (dstandish) [inline airflow/models/dagrun.py:653]: this is a drive-by.... it doesn't need to be RPC AFAICT

**Review by vincbeck — APPROVED** [review]

**Comment by dstandish** [comment]
this, dealing, as it does, with try_number, is a weird one, so i'm gonna let it go overnight to see if @uranusjr @potiuk @jedcunningham want to give it a look

**Review by jedcunningham — APPROVED** [review]
It all seems to make sense to me, but take my review with a grain of salt since 1) try number is confusing and 2) I'm not that familiar with our use of pydantic in this context.

**Review by hussein-awala — APPROVED** [review]

**Review by dirrao — APPROVED** [review]

File: airflow/models/dagrun.py (hunk 1 above)
Comment (uranusjr) [reply airflow/models/dagrun.py:653]: cc @vincbeck

File: airflow/models/taskinstance.py [hunk 2]
```diff
@@ -539,7 +539,7 @@ def _refresh_from_db(
         task_instance.end_date = ti.end_date
         task_instance.duration = ti.duration
         task_instance.state = ti.state
-        task_instance.try_number = ti._try_number  # private attr to get value unaltered by accessor
+        task_instance.try_number = _get_private_try_number(task_instance=ti)
```
Comment (uranusjr) [inline airflow/models/taskinstance.py:542]: I wonder if this can be improved when we implement TaskInstanceRun…

Comment (dstandish) [reply airflow/models/taskinstance.py:542]: Yes this problem should go away completely 

I actually doubt we really need these shenanigans now (specifically the conditional +1 logic)

File: airflow/models/dagrun.py (hunk 1 above)
Comment (dstandish) [reply airflow/models/dagrun.py:653]: Basis for conclusion is, I can run task without it.

File: airflow/models/taskinstance.py (hunk 2 above)
Comment (uranusjr) [reply airflow/models/taskinstance.py:542]: I don’t know either (never fully understood the logic tbh). Talking to Ash may help.

Comment (dstandish) [reply airflow/models/taskinstance.py:542]: According to the comments it is just “to ensure logs go to right file “ but if that’s really all, it could be corrected in the file task handler and try number could just be one value

**Review by potiuk — APPROVED** [review]

File: airflow/models/taskinstance.py (hunk 2 above)
Comment (potiuk) [reply airflow/models/taskinstance.py:542]: I think there were at least one or two attempts I remember in the past to remove it - and failed so far. 

I believe there is race conditions that try number is increased in the DB when task state is changed from RUNNING to FAILED (for example when user stops the task from the UI)  and try_number is already stored in the DB - but the running task might still be running and stopping itself, so it should continue logging to the same try_number it was started with.
